src/dwg.c

```diff
@@ -3192,3 +3205,230 @@ dwg_log_dataflags (const int _loglevel, const int maxlevel,
+/***
+ *void *bsearch_ex() - do a binary search on an array
+ *
+ *Purpose:
+ *   Does a binary search of a sorted array for a key.
+ *
+ *Entry:
+ *   const void *key    - key to search for
+ *   const void *base   - base of sorted array to search
+ *   size_t num         - number of elements in array
+ *   size_t width       - number of bytes per element
+ *   int (*compare)(const void* item1,const void* item2) - pointer to function
+ *that compares two array elements, returning neg when #1 < #2, pos when #1 >
+ *#2, and 0 when they are equal. Function is passed pointers to two array
+ *elements. const void** ppBefore - pointer to key's insert pos in array,if key
+ *not exist if *ppBefore = NULL,key should append at tail if ppBefore = NULL,
+ *caller don't care about insert pos.
+ *
+ *Exit:
+ *   if key is found:
+ *           returns pointer to occurrence of key in array
+ *           *ppBefore = NULL
+ *   if key is not found:
+ *           returns NULL
+ *           *ppBefore pointer to key's insert pos in array
+ *
+ *Exceptions:
+ *   Input parameters are validated. Refer to the validation section of the
+ *function.
+ *
+ *******************************************************************************/
+static void *
+bsearch_ex (const void *key, const void *base, size_t num, size_t width,
+            int (*compare) (const void *item1, const void *item2),
+            const void **ppBefore)
+{
+  if (NULL != ppBefore)
+    {
+      *ppBefore = NULL;
+    }
+  if (0 == num)
+    {
+      return NULL;
+    }
+  size_t num_now = num;
+  char *lo = (char *)base;
+  char *hi = (char *)base + (num_now - 1) * width;
+  for (; lo <= hi;)
+    {
+      if (0 == num_now)
+        {
+          break;
+        }
+      if (1 == num_now)
+        {
+          if (0 == compare (key, lo))
+            {
+              // key == lo
+              return lo;
+            }
+          break;
+        }
+      else
+        {
+          size_t half = num_now / 2;
+          char *mid = lo + (num_now & 1 ? half : (half - 1)) * width;
+          int result = compare (key, mid);
+          if (0 == result)
+            {
+              // key == mid
+              return mid;
+            }
+          if (result < 0)
+            {
+              // key < mid
+              hi = mid - width;
+              num_now = num_now & 1 ? half : half - 1;
+            }
+          else
+            {
+              // key > mid
+              lo = mid + width;
+              num_now = half;
+            }
+        }
+    }
+  if (NULL != ppBefore)
+    {
+      if (compare (key, lo) > 0)
+        {
+          // key > lo
+          lo += width;
+          if (lo < (char *)base + num * width)
+            {
+              *ppBefore = lo;
+            }
+        }
+      else
+        {
+          // key < lo
+          *ppBefore = lo;
+        }
+    }
+  return NULL;
+}
+
+void
+dwg_ordered_ref_appended (Dwg_Data *restrict dwg, Dwg_Object_Ref *ref)
```

Comment:
maybe just dwg_ordered_object_refs_add(). It's either insert or append

Reply:
OK.